### DISCUSSION TITLE
Hidden details of tractography

### DISCUSSION THREAD

**Discussion by EmmaRenauld** [dipy/dipy#2764] [Tractography / Tractometry] [answered]
Hi,

Could it be possible to add more details to the doc concerning the stopping criteria? Can you correct me if I am wrong below?

  - When a streamlines reaches a stopping criteria:
            - If it is because of the maximal angle, this last point is NOT added to the streamline*.
            - If is is because of other criteria depending on the choice of tractography (ex: threshold on the ODF), NO new direction is chosen and no new point is added to the streamline*.
            - If it is because of the tracking mask, this last point IS added to the streamline**.
    

Is this behaviour really what is expected by the community? If I am not wrong in my description above, then all streamlines can potentially end out of the tracking mask. We discussed it a little here in our lab, and we wonder if it would be more clear  for users to provide tracking masks that include both the white matter and the interface, and have streamlines entirely included in the tracking mask. Less dangerous for connectivity metrics, too.

Thank you!

<br>

<br>

Notes:

\* I understand this from direction getter's `get_direction_c` usage in  [dipy.tracking.direction_getter, Line120](https://github.com/dipy/dipy/blob/77afd6098c70939dde1647e71fa1d6c7be1ef48e/dipy/tracking/direction_getter.pyx#L120)   where the streamline propagation BREAKS when it returns a 1, i.e. when a new direction was not found (ex, see [dipy.direction.closest_peak_direction_getter, Line204](https://github.com/dipy/dipy/blob/77afd6098c70939dde1647e71fa1d6c7be1ef48e/dipy/direction/closest_peak_direction_getter.pyx#L204), where the angle is verified).

\** I understand this from [Line125](https://github.com/dipy/dipy/blob/77afd6098c70939dde1647e71fa1d6c7be1ef48e/dipy/tracking/direction_getter.pyx#L125) where the new point is copied BEFORE verifying the stream_status's stopping criteria, and THEN the generation breaks.

**Comment by arnaudbore** [comment]
@skoudoro @Garyfallidis ?
Thank you in advance!

**Reply by EmmaRenauld** [reply]
@gabknight

**Comment by gabknight** [comment, marked as answer]
Hi @EmmaRenauld, @arnaudbore,

We will clarify this in the doc for the next release. The last points are always saved. 

```
When a streamlines reaches a stopping criteria:
- If it is because of the maximal angle, this last point is NOT added to the streamline*.
```
No, if there is no direction available to follow in the opening cone, the tracking stops, and the current point is included in the streamline.

`- If is is because of other criteria depending on the choice of tractography (ex: threshold on the ODF), NO new direction is chosen and no new point is added to the streamline*.`
Same as for the maximal angle, if there is no direction available to follow, the tracking stops, the current point is included in the streamline.

`- If it is because of the tracking mask, this last point IS added to the streamline**.`
no, if the streamline stops because of the stopping criteria object (e.g. stepping outside a binary mask), the point that trigger this event IS NOT included in the streamline.

`Is this behaviour really what is expected by the community? If I am not wrong in my description above, then all streamlines can potentially end out of the tracking mask. We discussed it a little here in our lab, and we wonder if it would be more clear for users to provide tracking masks that include both the white matter and the interface, and have streamlines entirely included in the tracking mask. Less dangerous for connectivity metrics, too.`

The documentation will be improved in the to make this behavior clear.  If tractography stops because of the SC, the point that triggers is not kept. If tractography stops because of the orientation data (ODF, angle), the tractography stops at the furthest point it can go. If the tractography reaches outside the image volume, the tractography will stop and that point will be excluded.

[Edited, the point triggering the stopping criteria *IS NOT* included in the streamline since PR #1666, an option should be added as suggested in #1886]

**Comment by EmmaRenauld** [comment]
> No, if there is no direction available to follow in the opening cone, the tracking stops, and the current point is included in the streamline.

To clarify, the point *after* the wrong angle is *not* kept?

If so, ok, @arnaudbore , after my PR, we will have the same in scilpy.

**Reply by gabknight** [reply]
yes, there is no point after. The possible tracking directions at the current position are estimated. If there is none valid (e.g. within the opening angle and ODF > threshold), the current tracking position is included, but the tracking stops.

**Comment by gabknight** [comment]
I corrected my mistake above. @EmmaRenauld  @arnaudbore the point triggering the stopping criteria **IS NOT** included in the streamline since PR #1666, an option should be added as suggested in #1886; and better documentation.

If the streamline reaches an `ENDPOINT` or an `OUTSIDEIMAGE` status, that point **is not included** in the final tractogram. i.e. if you track within a binary mask (`BinaryStoppingCriterion`) no streamlines points should be outside this mask.

**Comment by EmmaRenauld** [comment]
@arnaudbore Then we should discuss with the lab if our option should be set to false by default.
